modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/EnterpriseGeoIpDownloader.java

```diff
@@ -0,0 +1,449 @@
+        String expectedChecksum,
+        long timestamp
+    ) throws IOException {
+        MessageDigest md5 = MessageDigests.md5();
+        for (byte[] buf = getChunk(is); buf.length != 0; buf = getChunk(is)) {
+            md5.update(buf);
+            if (digest != null) {
+                digest.update(buf);
+            }
+            IndexRequest indexRequest = new IndexRequest(DATABASES_INDEX).id(name + "_" + chunk + "_" + timestamp)
+                .create(true)
+                .source(XContentType.SMILE, "name", name, "chunk", chunk, "data", buf);
+            client.index(indexRequest).actionGet();
+            chunk++;
+        }
+
+        // May take some time before automatic flush kicks in:
+        // (otherwise the translog will contain large documents for some time without good reason)
+        FlushRequest flushRequest = new FlushRequest(DATABASES_INDEX);
+        client.admin().indices().flush(flushRequest).actionGet();
+        // Ensure that the chunk documents are visible:
+        RefreshRequest refreshRequest = new RefreshRequest(DATABASES_INDEX);
+        client.admin().indices().refresh(refreshRequest).actionGet();
+
+        String actualMd5 = MessageDigests.toHexString(md5.digest());
+        String actualChecksum = digest == null ? actualMd5 : MessageDigests.toHexString(digest.digest());
+        if (Objects.equals(expectedChecksum, actualChecksum) == false) {
+            throw new IOException("checksum mismatch, expected [" + expectedChecksum + "], actual [" + actualChecksum + "]");
+        }
+        return Tuple.tuple(chunk, actualMd5);
+    }
+
+    // visible for testing
+    static byte[] getChunk(InputStream is) throws IOException {
+        byte[] buf = new byte[MAX_CHUNK_SIZE];
+        int chunkSize = 0;
+        while (chunkSize < MAX_CHUNK_SIZE) {
+            int read = is.read(buf, chunkSize, MAX_CHUNK_SIZE - chunkSize);
+            if (read == -1) {
+                break;
+            }
+            chunkSize += read;
+        }
+        if (chunkSize < MAX_CHUNK_SIZE) {
+            buf = Arrays.copyOf(buf, chunkSize);
+        }
+        return buf;
+    }
+
+    /**
+     * Downloads the geoip databases now, and schedules them to be downloaded again after pollInterval.
+     */
+    synchronized void runDownloader() {
+        // by the time we reach here, the state will never be null
+        assert state != null;
+
+        // there's a race condition between here and requestReschedule. originally this scheduleNextRun call was at the end of this
+        // block, but remember that updateDatabases can take seconds to run (it's downloading bytes from the internet), and so during the
+        // very first run there would be no future run scheduled to reschedule in requestReschedule. which meant that if you went from zero
+        // to N(>=2) databases in quick succession, then the all but the first database wouldn't necessarily get downloaded, because the
```

Comment:
Random typo (I think):

```suggestion
        // to N(>=2) databases in quick succession, then all but the first database wouldn't necessarily get downloaded, because the
```